### ISSUE TITLE
ROI-Focused Autoregressive Generation with a Dual-Decoder Architecture

### ISSUE THREAD

**Issue by MartialTerran** [NASA-IMPACT/Surya#31]
### **New Issue: ROI-Focused Autoregressive Generation with a Dual-Decoder Architecture**

**Opened by:** Martial Terran mrferran1970@gmail.com  https://huggingface.co/MartialTerran
**Date:** 2025-09-11

**Labels:** `enhancement`, `autoregressive`, `forecasting`, `ROI`, `diffusion-model`

### **1. Summary**

This proposal outlines a plan to enhance the Surya model with a Region of Interest (ROI) focused autoregressive generation capability. This will be achieved by implementing a dual-decoder architecture with a second diffusion head specifically trained to predict a smaller subframe centered on a designated ROI (e.g., a sunspot of interest). This will allow for more accurate and detailed predictions of solar phenomena within the specified region while maintaining the global context of the full solar disk.

### **2. Background and Motivation**

The previously proposed autoregressive enhancement to Surya allows for extended solar forecasting. However, to further increase prediction accuracy, particularly for dynamic phenomena like solar flares which originate from specific active regions, a more focused approach is needed. By concentrating the model's predictive power on a user-defined ROI, we can achieve higher fidelity predictions for that specific area.

This will be accomplished by adding a second decoder and diffusion head to the existing Surya model. The first decoder will continue to predict the full solar disk, while the second will be trained to predict a smaller, high-resolution subframe centered on the ROI. This approach is inspired by recent work in controllable image generation and multi-headed diffusion models.

### **3. Proposed Engineering Plan**

**Phase 1: Dual-Decoder Architecture and ROI Subframe Tokenization**

1.  **Subframe Token Input:**
    *   The ROI will be defined by a bounding box, and the corresponding subframe will be extracted and tokenized separately from the full-frame image.
    *   These subframe tokens will be concatenated with the full-frame tokens and fed into the model's shared encoder. A special "ROI token" will be used to delineate the subframe tokens from the full-frame tokens.

2.  **Dual-Decoder and Diffusion Heads:**
    *   The output of the shared encoder will be passed to two separate decoders.
    *   **Decoder 1 (Full-Frame):** This will be the existing Surya decoder, which will be trained on the full-frame image. The output of this decoder will be passed to the existing diffusion head to generate the full-frame prediction.
    *   **Decoder 2 (Subframe):** A new, smaller decoder will be added and trained specifically on the subframe tokens. The output of this decoder will be passed to a new, second diffusion head to generate the high-resolution subframe prediction.

**Phase 2: Training the Dual-Decoder Model**

1.  **Combined Loss Function:** The model will be trained with a combined loss function that includes the loss from both the full-frame and subframe predictions. The relative weights of the two losses can be adjusted to prioritize the accuracy of either the full-frame or subframe prediction.

2.  **Staged Training:** To ensure stable training, we will employ a staged approach:
    *   **Stage 1:** Freeze the weights of the pre-trained full-frame model and train only the new subframe decoder and diffusion head.
    *   **Stage 2:** Unfreeze all weights and fine-tune the entire model end-to-end.

**Phase 3: Autoregressive Rollout with ROI Focus**

1.  **ROI-Centered Rollout:** The autoregressive rollout script will be modified to incorporate the dual-decoder architecture. At each step, the model will generate both a full-frame and a subframe prediction. The subframe prediction will then be "pasted" back onto the full-frame prediction at the ROI location to create a more accurate input for the next time step.

### **4. Code Snippets and Implementation Details**

The following pseudo-code illustrates the proposed dual-decoder architecture and training process:

```python
import torch
from surya.model import SuryaModel, Encoder, Decoder, DiffusionHead

class SuryaROIDecoder(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.encoder = Encoder()
        self.full_frame_decoder = Decoder()
        self.sub_frame_decoder = Decoder() # Potentially smaller decoder
        self.full_frame_diffusion_head = DiffusionHead()
        self.sub_frame_diffusion_head = DiffusionHead()

    def forward(self, full_frame_tokens, sub_frame_tokens):
        # Concatenate tokens and pass through encoder
        combined_tokens = torch.cat([full_frame_tokens, sub_frame_tokens], dim=1)
        encoded_features = self.encoder(combined_tokens)

        # Split features for decoders
        full_frame_features = encoded_features[:, :full_frame_tokens.shape[1], :]
        sub_frame_features = encoded_features[:, full_frame_tokens.shape[1]:, :]

        # Decode and generate predictions
        full_frame_pred = self.full_frame_diffusion_head(self.full_frame_decoder(full_frame_features))
        sub_frame_pred = self.sub_frame_diffusion_head(self.sub_frame_decoder(sub_frame_features))

        return full_frame_pred, sub_frame_pred

# Example Training Loop
model = SuryaROIDecoder()
optimizer = torch.optim.Adam(model.parameters())
loss_fn_full = torch.nn.MSELoss()
loss_fn_sub = torch.nn.MSELoss()

for epoch in range(num_epochs):
    for full_frame_batch, sub_frame_batch, roi_coords in train_dataloader:
        optimizer.zero_grad()

        # Get tokens for full and sub frames
        full_frame_tokens = tokenize(full_frame_batch)
        sub_frame_tokens = tokenize(sub_frame_batch)

        # Forward pass
        full_frame_pred, sub_frame_pred = model(full_frame_tokens, sub_frame_tokens)

        # Calculate loss
        loss_full = loss_fn_full(full_frame_pred, full_frame_batch)
        loss_sub = loss_fn_sub(sub_frame_pred, sub_frame_batch)
        total_loss = loss_full + loss_sub # Potentially weighted

        # Backward pass and optimization
        total_loss.backward()
        optimizer.step()```

### **5. Evaluation and Success Metrics**

The success of this project will be measured by:

*   **Improved ROI Prediction Accuracy:** A significant improvement in the prediction accuracy within the ROI compared to the full-frame model.
*   **Reduced Error Accumulation:** A reduction in the accumulation of errors within the ROI during the autoregressive rollout.
*   **Qualitative Assessment:** A qualitative assessment of the generated subframes to ensure they are physically plausible and coherent.

### **6. Conclusion**

The proposed dual-decoder architecture with a dedicated ROI diffusion head represents a significant step forward in our ability to accurately forecast solar phenomena. By focusing the model's predictive power on specific regions of interest, we can achieve higher fidelity predictions and extend the useful range of our solar forecasts. We welcome feedback and contributions from the community on this proposal.

[1]  https://www.researchgate.net/figure/Diffusion-Model-Architecture-During-training-our-two-headed-model-U-left-learns-to_fig4_390922058 
[2] https://www.researchgate.net/figure/Diffusion-Model-Architecture-During-training-our-two-headed-model-U-left-learns-to_fig1_389547819 
[ 3] Improving Training Efficiency of Diffusion Models via Multi-Stage Framework and Tailored Multi-Decoder Architecture  https://openaccess.thecvf.com/content/CVPR2024/papers/Zhang_Improving_Training_Efficiency_of_Diffusion_Models_via_Multi-Stage_Framework_and_CVPR_2024_paper.pdf
